Convert to vite build and esm

siefkenj opened the pull request:
This PR adds the Vite build system, which assembles the Javascript into a `dist/` directory. It also switches the project to `type: "module"`.

The Javascript was minimally modified to conform to es modules. This is in preparation for a conversion to Typescript, but I wanted to get a build system working before majorly changing the source of any files.

You can build the files with `npm run build`. This needs to be done prior to running `npm run test` since the `test` scrip imports directly from the built Javascript.

This PR adds an unused dependency to `vitest`. If this PR is in the right direction, I will convert the tests to `vitest`.

Questions/Comments:
 * There is an `.editorconfig` file that specifies 2 spaces for indentation and a conflicting `.prettierrc` that specifies 4 spaces. Which is preferred?
 * Why are so many files included in the `files` list for the package when published to NPM? `README.md` is automatically included, and so doesn't need to be listed. I assume people don't generally expect test scripts to be installed when they do `npm install ts-pegjs`... I cut those includes down to just things needed to run the project.

pjmolina commented:
Hi. Thanks for the PR.
Looks good path.
One minor thing: `npm run test:cli` is failing due to change to modules. Can we fix this one?

About your questions:
1. 2 spaces will be fine, thanks.
2. Makes sense. +1

siefkenj commented:
I have fixed `test:cli` and added a Github Action script that should run all test on a PR; you'll have to enable actions, though.

Since `peggy` was `require()`ing tspegjs, I converted the package back to commonjs with dual delivery. That is, both `.mjs` es modules and `.js` commonjs modules are built. The `"exports"` line in `package.json` should make sure a user imports the appropriate library when the use tspegjs. Hopefully there won't be further issues :-).

siefkenj commented:
@pjmolina I have converted the tests to `vitest`. They take a bit longer to run now because `tsc` is called once per generated file rather than once for all files together. I can change it back to the old behavior if you want, but at the moment it has the benefit that it will say which file encounters an error.

Also by running `npx vitest` `vitest` will run in watch mode and re-run all tests upon any code change. Hopefully this sets things up for easy refactoring.

 1. I noticed the old `test:cli` didn't actually test the CLI but instead tested loading `tspegjs` as a plugin. Was this an intentional omission? 
 2. I also noticed that `tslint` is not run by default and throws many errors. Should it just be removed?

pjmolina commented on tsconfig.json:
```diff
@@ -0,0 +1,21 @@
+    "resolveJsonModule": true,
+    "isolatedModules": true,
+    "noEmit": true,
+    "jsx": "react"
```
jsx ? not needed, right?

pjmolina commented:
1. The tests under `test:cli` are important due to tests the output of the parser with different settings but also compiles to TS to JS and pass the linter (non only to `src/` but to the output samples to ensure coverage). I see all of this missing in your PR:
```
  "test": "npm run gen:samples && npm run compile:samples && npm run lint:samples && npm run lint && npm run test:samples",
        "test:es": "npm run gen:es && npm run compile:samples && npm run lint:samples && npm run lint && npm run test:samples",
        "test:non-optimized": "npm run gen:samples && npm run compile:samples && npm run lint:samples && npm run lint && npm run test:samples",
        "test:optimize-size": "npm run gen:samples-optimize-size && npm run compile:samples && npm run lint:samples && npm run lint && npm run test:samples",
        "test:samples": "node test/test-samples.js",
        "gen:es": "node test/test.js --cache --trace --format es",
        "gen:samples": "node test/test.js --cache --trace",
        "gen:samples-optimize-size": "node test/test.js --cache --trace --optimize size",
        "compile:samples": "tsc --target es6 --module commonjs --declaration output/arithmetics.ts output/css.ts output/json.ts output/javascript.ts output/st.ts output/arithmetics-typed.ts output/bulkOpening.ts output/minimal-typed.ts ",
```
2. Tes tslint is deprecated and can be removed in favour of eslint.

pjmolina commented:
Still failing:
```
λ npm test

> ts-pegjs@3.1.0 pretest
> npm run clean


> ts-pegjs@3.1.0 clean
> rimraf output/*.ts output/*.js


> ts-pegjs@3.1.0 test
> vitest run


 RUN  v0.29.7 C:/dev/ts-pegjs-siefkenj

 ❯ test/cli.test.ts (3) 9806ms
   ❯ CLI Tests (3) 9805ms
     ✓ Can import tspegjs as a Peggy plugin 4745ms
     × Generated `ts` file passes eslint check 5055ms
     × Can compile `ts` file to `js` 5057ms
 ❯ test/generate.test.ts (9) 25242ms
   ❯ Can generate parser for `arithmetics.pegjs` with custom return type (5) 15145ms
     ✓ Can generate parser for `arithmetics.pegjs`
     × Generated `ts` file passes eslint check 5042ms
     ✓ Generated `ts` file contains custom header
     × Can compile `ts` file to `js` 5043ms
     × Can parse arithmetics 5022ms
   ❯ Can generate parser for `minimal.pegjs` with custom return type (4) 10095ms
     ✓ Can generate parser for `minimal.pegjs`
     × Generated `ts` file passes eslint check 5054ms
     ✓ Generated `ts` file contains custom header
     × Can compile `ts` file to `js` 5029ms
 ❯ test/arithmetic.test.ts (32) 42544ms
   ❯ Build and lint samples (32) 42542ms
     ❯ arithmetics.pegjs (4) 5105ms
       ✓ Can generate parser for `arithmetics.pegjs`
       × Generated `ts` file passes eslint check 5036ms
       ✓ Generated `ts` file contains custom header
       × Can compile `ts` file to `js` 5063ms
     ❯ json.pegjs (4) 5135ms
       ✓ Can generate parser for `json.pegjs`
       × Generated `ts` file passes eslint check 5037ms
       ✓ Generated `ts` file contains custom header
       × Can compile `ts` file to `js` 5076ms
     ❯ css.pegjs (4) 5593ms
       ✓ Can generate parser for `css.pegjs` 343ms
       × Generated `ts` file passes eslint check 5073ms
       ✓ Generated `ts` file contains custom header 367ms
       × Can compile `ts` file to `js` 5246ms
     ❯ javascript.pegjs (4) 6173ms
       ✓ Can generate parser for `javascript.pegjs` 1098ms
       × Generated `ts` file passes eslint check 5047ms
       ✓ Generated `ts` file contains custom header
       × Can compile `ts` file to `js` 5072ms
     ❯ st.pegjs (4) 5285ms
       ✓ Can generate parser for `st.pegjs`
       × Generated `ts` file passes eslint check 5035ms
       ✓ Generated `ts` file contains custom header
       × Can compile `ts` file to `js` 5057ms
     ❯ bulkOpening.pegjs (4) 5103ms
       ✓ Can generate parser for `bulkOpening.pegjs`
       × Generated `ts` file passes eslint check 5030ms
       ✓ Generated `ts` file contains custom header
       × Can compile `ts` file to `js` 5068ms
     ❯ minimal.pegjs (4) 5064ms
       ✓ Can generate parser for `minimal.pegjs`
       × Generated `ts` file passes eslint check 5034ms
       ✓ Generated `ts` file contains custom header
       × Can compile `ts` file to `js` 5060ms
     ❯ Can generate parser for `minimal.pegjs` with custom return type (4) 5081ms
       ✓ Can generate parser for `minimal.pegjs`
       ✓ Generated `ts` file passes eslint check 4898ms
       ✓ Generated `ts` file contains custom header
       × Can compile `ts` file to `js` 5072ms

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Failed Tests 22 ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯

 FAIL  test/arithmetic.test.ts > Build and lint samples > arithmetics.pegjs > Generated `ts` file passes eslint check
 FAIL  test/arithmetic.test.ts > Build and lint samples > arithmetics.pegjs > Can compile `ts` file to `js`
 FAIL  test/arithmetic.test.ts > Build and lint samples > json.pegjs > Generated `ts` file passes eslint check
 FAIL  test/arithmetic.test.ts > Build and lint samples > json.pegjs > Can compile `ts` file to `js`
 FAIL  test/arithmetic.test.ts > Build and lint samples > css.pegjs > Generated `ts` file passes eslint check
 FAIL  test/arithmetic.test.ts > Build and lint samples > css.pegjs > Can compile `ts` file to `js`
 FAIL  test/arithmetic.test.ts > Build and lint samples > javascript.pegjs > Generated `ts` file passes eslint check
 FAIL  test/arithmetic.test.ts > Build and lint samples > javascript.pegjs > Can compile `ts` file to `js`
 FAIL  test/arithmetic.test.ts > Build and lint samples > st.pegjs > Generated `ts` file passes eslint check
 FAIL  test/arithmetic.test.ts > Build and lint samples > st.pegjs > Can compile `ts` file to `js`
 FAIL  test/arithmetic.test.ts > Build and lint samples > bulkOpening.pegjs > Generated `ts` file passes eslint check
 FAIL  test/arithmetic.test.ts > Build and lint samples > bulkOpening.pegjs > Can compile `ts` file to `js`
 FAIL  test/arithmetic.test.ts > Build and lint samples > minimal.pegjs > Generated `ts` file passes eslint check
 FAIL  test/arithmetic.test.ts > Build and lint samples > minimal.pegjs > Can compile `ts` file to `js`
 FAIL  test/arithmetic.test.ts > Build and lint samples > Can generate parser for `minimal.pegjs` with custom return type > Can compile `ts` file to `js`
 FAIL  test/cli.test.ts > CLI Tests > Generated `ts` file passes eslint check
 FAIL  test/cli.test.ts > CLI Tests > Can compile `ts` file to `js`
 FAIL  test/generate.test.ts > Can generate parser for `arithmetics.pegjs` with custom return type > Generated `ts` file passes eslint check
 FAIL  test/generate.test.ts > Can generate parser for `arithmetics.pegjs` with custom return type > Can compile `ts` file to `js`
 FAIL  test/generate.test.ts > Can generate parser for `arithmetics.pegjs` with custom return type > Can parse arithmetics
 FAIL  test/generate.test.ts > Can generate parser for `minimal.pegjs` with custom return type > Generated `ts` file passes eslint check
 FAIL  test/generate.test.ts > Can generate parser for `minimal.pegjs` with custom return type > Can compile `ts` file to `js`
Error: Test timed out in 5000ms.
If this is a long-running test, pass a timeout value as the last argument or configure it globally with "testTimeout".
⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[1/22]⎯

 Test Files  3 failed (3)
      Tests  22 failed | 22 passed (44)
   Start at  17:51:27
   Duration  44.41s (transform 4.97s, setup 0ms, collect 1.30s, tests 77.59s)
```

siefkenj commented:
Looks like it is now just typescript taking a long time. I've increased the test timeout to 20s (up from 5).

@pjmolina Can you enable the GH Actions workflow for this PR? Then we will also see the results of the tests running on GH's CI.

pjmolina commented:
The timeout improved it. Still 2 tests failing:
```

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Failed Tests 2 ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯

 FAIL  test/generate.test.ts > Can generate parser for `arithmetics.pegjs` with custom return type > Generated `ts` file passes eslint check
Error: Command failed: eslint 'C:\dev\ts-pegjs-siefkenj\output\arithmetics-typed.ts'

Oops! Something went wrong! :(

ESLint: 8.35.0

No files matching the pattern "'C:\dev\ts-pegjs-siefkenj\output\arithmetics-typed.ts'" were found.
Please check for typing mistakes in the pattern.


 ❯ ChildProcess.exithandler node:child_process:419:12
 ❯ ChildProcess.emit node:events:513:28
 ❯ Socket.emit node:events:513:28
 ❯ Pipe.<anonymous> node:net:322:12

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯Serialized Error: {
  "cmd": "eslint 'C:\dev\ts-pegjs-siefkenj\output\arithmetics-typed.ts'",
  "code": 2,
  "killed": false,
  "signal": null,
  "stderr": "
Oops! Something went wrong! :(

ESLint: 8.35.0

No files matching the pattern "'C:\dev\ts-pegjs-siefkenj\output\arithmetics-typed.ts'" were found.
Please check for typing mistakes in the pattern.

",
  "stdout": "",
}
⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[1/2]⎯

 FAIL  test/generate.test.ts > Can generate parser for `arithmetics.pegjs` with custom return type > Can compile `ts` file to `js`
Error: Command failed: tsc --target es6 --module commonjs --declaration 'C:\dev\ts-pegjs-siefkenj\output\arithmetics-typed.ts'

 ❯ ChildProcess.exithandler node:child_process:419:12
 ❯ ChildProcess.emit node:events:513:28

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯Serialized Error: {
  "cmd": "tsc --target es6 --module commonjs --declaration 'C:\dev\ts-pegjs-siefkenj\output\arithmetics-typed.ts'",
  "code": 2,
  "killed": false,
  "signal": null,
  "stderr": "",
  "stdout": "error TS6054: File ''C:/dev/ts-pegjs-siefkenj/output/arithmetics-typed.ts'' has an unsupported extension. The only supported extensions are '.ts', '.tsx', '.d.ts', '.cts', '.d.cts', '.mts', '.d.mts'.
  The file is in the program because:
    Root file specified for compilation
",
}
⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[2/2]⎯

 Test Files  1 failed | 2 passed (3)
      Tests  2 failed | 42 passed (44)
   Start at  18:10:06
   Duration  69.19s (transform 853ms, setup 0ms, collect 1.27s, tests 110.39s)
```

pjmolina commented:
All good now!! 
Ready to merge?

siefkenj commented:
> All good now!! Ready to merge?

Looks good to me!

pjmolina commented:
Merged!

pjmolina commented:
test:cli fails after merging: https://github.com/metadevpro/ts-pegjs/actions/runs/4534621830/jobs/7989040845

siefkenj commented:
> test:cli fails after merging: https://github.com/metadevpro/ts-pegjs/actions/runs/4534621830/jobs/7989040845.
> 
> Adding build as a prerequiste for test

Thanks for fixing that!
